src/Chainweb/Payload/PayloadStore.hs

```diff
@@ -343,29 +291,29 @@ instance PayloadCasLookup_ a cas => HasCasLookup (PayloadDb_ a cas) where
             , _payloadWithOutputsTransactionsHash = txsHash
             , _payloadWithOutputsOutputsHash = outsHash
             }
-    {-# INLINE casLookup #-}
+    {-# INLINE tableLookup #-}
 
 
 -- | Combine all Payload related stores into a single content addressed
 -- store. We want the invariant that if a key is present in the store also all
 -- of its dependencies are present. For that we must be careful about the order
 -- of insertion and deletions.
 --
-instance (MerkleHashAlgorithm a, PayloadCas_ a cas) => IsCas (PayloadDb_ a cas) where
-    casInsert = addNewPayload
-    {-# INLINE casInsert #-}
+instance (pk ~ CasKeyType (PayloadWithOutputs_ a), MerkleHashAlgorithm a, CanPayloadCas_ a tbl) => Table (PayloadDb_ a tbl) pk (PayloadWithOutputs_ a) where
+    tableInsert db _ v = addNewPayload db v
+    {-# INLINE tableInsert #-}
 
-    casDelete db k =
-        casLookup (_transactionDbBlockPayloads $ _transactionDb db) k >>= \case
+    tableDelete db k = 
+        tableLookup (_transactionDbBlockPayloads $ _transactionDb db) k >>= \case
             Just pd -> do
-                casDelete
+                tableDelete
                     (_transactionDbBlockPayloads $ _transactionDb db)
                     k
-                casDelete
+                tableDelete
                     (_transactionDbBlockTransactions $ _transactionDb db)
                     (_blockPayloadTransactionsHash pd)
-                casDelete
+                tableDelete
                     (_payloadCacheBlockOutputs $ _payloadCache db)
                     (_blockPayloadOutputsHash pd)
             Nothing -> return ()
-    {-# INLINE casDelete #-}
+    {-# INLINE tableDelete #-}
```

Comment:
You'd think we could just write
```hs
instance (MerkleHashAlgorithm a, CanPayloadCas_ a tbl) => 
  Cas (PayloadDb_ a tbl) (PayloadWithOutputs_ a)
```
since `type Cas t v = Table t (CasKeyType v) v`. Unfortunately, that instance is not possible to write due to https://gitlab.haskell.org/ghc/ghc/-/issues/13267.

You might think that instead we can write
```hs
instance (MerkleHashAlgorithm a, CanPayloadCas_ a tbl) => 
  Table (PayloadDb_ a tbl) (CasKeyType (PayloadWithOutputs_ a)) (PayloadWithOutputs_ a)
```
but this is illegal in the first place, this time deliberately, see https://gitlab.haskell.org/ghc/ghc/-/issues/13262 or https://gitlab.haskell.org/ghc/ghc/-/issues/14046. We can't have type family applications in our instance heads at all. So instead, we need to ask for `k ~ CasKeyType (PayloadOutputs_ a)`, and in the instance head we ask for a `Table` with key type `k`. This pattern repeats in other places in this file.

Reply:
So, the definitions of `ReadableTable` and `Table` are:
``` haskell
class ReadableTable t k v | t -> k v where
  ...
class ReadableTable t k v => Table t k v | t -> k v where
  ...
```
If we instead had:
``` haskell
class ReadableTable t where
  type TableKey t
  type TableValue t 
  ...
class ReadableTable t => Table t where  
  ...
```

Then these instance heads wouldn't be nearly as gnarly:
```
instance (MerkleHashAlgorithm a, CanReadablePayloadCas_ a tbl) => ReadableTable (PayloadDb_ a tbl) 
  type TableKey (PayloadDb_ a tbl) = CasKeyType (PayloadWithOutputs_ a)
  type TableValue (PayloadDb_ a tbl) = PayloadWithOutputs_ a
  ...
instance (MerkleHashAlgorithm a, CanPayloadCas_ a tbl) => Table (PayloadDb_ a tbl) 
  -- No need to repeat the `TableKey` & `TableValue` types
  ...
```

I'm not sure if it would be useful here, but using associated types of a single class parameter instead of encoding them as functional dependencies between multiple parameters also enables us to use `DerivingVia`.

So, is there any reason why we're using functional dependencies for `ReadableTable` and `Table`?

Reply:
I've forgotten whether I tried that, so I'm going to give it a try now.

Reply:
Ah so after trying this, the places that don't need to talk about (constrain) the `TableKey` and `TableValue` types are shorter, but most places that *do* need to constrain them is way more verbose, with a bunch of equality constraints.

More generally the issue is, say we have `class C x y | x -> y`. Whenever we don't need to constrain `y`, the mention of `y` in constraints is useless verbosity. But when we do need to constrain it, say that we want it to equal a ground type `T`, we can just write `C x T` as a constraint, whereas with tyfams we have to write `(C x, CTyFam x ~ T)`, and this is only more verbose with more tyfams/dependent type parameters. And that ends up really annoying in practice because almost *everyone* who wants to use a table is going to care about the key and value types being exactly equal to some ground type.

Reply:
Oh, that makes sense. Thanks for trying this out! BTW, this observation makes me wonder how much it would help to have the following type alias (constraint alias, really :) ):

``` haskell
type C' x y = (C x, CTyFam x ~ y)
```

Reply:
Assuming it works it'd help a bit but it doesn't make instances much less verbose, because the type families still have to be defined for the instance itself, and then we end up with both `C` and `C'` which is a bit much. I'm not sure it's worse, but it doesn't seem clearly better.